src/main/java/org/opensearch/knn/index/codec/KNN990Codec/NativeEngines990KnnVectorsWriter.java

```diff
@@ -0,0 +1,115 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ *
+ * Modifications Copyright OpenSearch Contributors. See
+ * GitHub history for details.
+ */
+
+package org.opensearch.knn.index.codec.KNN990Codec;
+
+import lombok.RequiredArgsConstructor;
+import org.apache.lucene.codecs.KnnFieldVectorsWriter;
+import org.apache.lucene.codecs.KnnVectorsWriter;
+import org.apache.lucene.codecs.hnsw.FlatVectorsWriter;
+import org.apache.lucene.index.FieldInfo;
+import org.apache.lucene.index.MergeState;
+import org.apache.lucene.index.SegmentWriteState;
+import org.apache.lucene.index.Sorter;
+import org.apache.lucene.util.IOUtils;
+import org.apache.lucene.util.RamUsageEstimator;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * A KNNVectorsWriter class for writing the vector data strcutures and flat vectors for Native Engines.
+ */
+@RequiredArgsConstructor
+public class NativeEngines990KnnVectorsWriter extends KnnVectorsWriter {
+    private static final long SHALLOW_SIZE = RamUsageEstimator.shallowSizeOfInstance(NativeEngines990KnnVectorsWriter.class);
+    private final SegmentWriteState segmentWriteState;
+    private final FlatVectorsWriter flatVectorsWriter;
+    private final List<NativeEngineVectorFieldsWriter<?>> fields = new ArrayList<>();
+    private boolean finished;
+
+    /**
+     * Add new field for indexing.
+     * In Lucene, we use single file for all the vector fields so here we need to see how we are going to make things
+     * work.
+     * @param fieldInfo {@link FieldInfo}
+     */
+    @Override
+    public KnnFieldVectorsWriter<?> addField(final FieldInfo fieldInfo) throws IOException {
+        final NativeEngineVectorFieldsWriter<?> newField = NativeEngineVectorFieldsWriter.create(fieldInfo, segmentWriteState.infoStream);
+        // TODO: we can build the graph here too iteratively. but right now I am skipping that as we need iterative
+        // graph build support on the JNI layer.
+        fields.add(newField);
+        return flatVectorsWriter.addField(fieldInfo, newField);
+    }
+
+    /**
+     * Flush all buffered data on disk. This is not fsync. This is lucene flush.
+     *
+     * @param maxDoc int
+     * @param sortMap {@link Sorter.DocMap}
+     */
+    @Override
+    public void flush(int maxDoc, final Sorter.DocMap sortMap) throws IOException {
+        // simply write data in the flat file
+        flatVectorsWriter.flush(maxDoc, sortMap);
+        // TODO: add code for creating Vector datastructures during lucene flush operation
```

Comment:
Oh so this doesnt actually create the native indices, right?

Reply:
No its not creating any native index. Right now this is just creating flat vectors file.

Reply:
Once the 2.16 is released I will start working on those changes. Keeping the PRs small to ensure that we can review the code in detail.